christinee.liu

chrisliux opened the pull request:
I’ve added usecases for Function & Comparator.

**Function**

**_FunctionView_**: Shows the UI where users can choose actions like "Current Price" or "Compare," and sends those choices to the controller.
**_FunctionViewModel_**: Manages the data needed for the view, like button labels, and helps keep track of the view state.
**_FunctionController_**: Handles user actions from the view and sends them to the interactor, deciding what happens next based on what the user selects.
**_FunctionPresenter_**: Takes output from the interactor, updates the view model, and makes sure the right screen is displayed.
**_FunctionInteractor_**: Contains the main logic for handling user actions which are switching to the "Current Price" and "Comparator" screens.
Input and Output Boundaries: Interfaces that make sure communication between different parts of the system is clear and organized.
**_FunctionFactory_**: Sets up and connects all the components of the function feature (view, view model, controller, interactor, and presenter) to ensure everything works together smoothly following clean architecture principles.



**Comparator**

**_ComparatorView_**: Displays the UI where users can compare properties. It shows labels for properties and buttons for actions like "Estimate Price," "Add New Property," and "User Profile," and sends user choices to the controller.
**_ComparatorViewModel_**: Manages the data needed for the view, such as property labels, button names, and login status, to keep the view state updated.
_ComparatorController_: Handles user actions from the view, such as navigating to the current price screen or checking login status, and sends these actions to the interactor.
ComparatorPresenter: Takes output from the interactor, updates the view model, and makes sure the right screen is displayed.
**_ComparatorInteractor_**: Contains the main logic for handling user actions, like navigating to the "Current Price," "Input," or "User Profile" screens, and then communicates with the presenter.
**_ComparatorState_**: Keeps track of other properties lists, and whether the user is logged in. **The data we get from the api model will we stored here!**
**_ComparatorFactory_**: Sets up and connects all the components, ensuring that they work together properly. It follows clean architecture principles to construct the app's UI flow by wiring together the view, controller, interactor, and presenter.
**_Input and Output Boundaries_**: Interfaces that ensure clear communication between different parts of the system for actions and updates.


From the **ApplicationState** Hadi worked on, it has the functionality to track if a user is logged in or just browsing as a guest. If a user is logged in, the top panel in Comparator shows a "User Profile" button.


Also, I’ve added unit tests for FunctionInteractor and ComparatorInteractor and managed to get 100% coverage.

Side note: I accidentally deleted LoginPresenter on my end, so I had to restore it from the main branch. That’s why it shows as modified in my recent changes, even though no actual modifications were made to its functionality.

chrisliux committed:
add function use case and function view

chrisliux committed:
make currentPrice and Comparator for user and guest be on the same view

chrisliux committed:
add Comparator's view, usecase, ui, and adapters

chrisliux committed:
remove unused function usecase's InputData, OutputData, UserDataAccessInterface.

chrisliux committed:
remove unused execute in function usecase.

chrisliux committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into christinee.liu

chrisliux committed:
add ComparatorFactory & FunctionFactory

chrisliux committed:
FEATURE: Implement ComparatorView & FunctionView.

chrisliux committed:
FEATURE: In order to have ApplicationState for login management, added checkLoginStatus() & goToUserProfile() functions in ComparatorController, along with adding updateLoginStatus() in ComparatorOutputBoundary & checkLoginStatus() in ComparatorInteractor.

FEATURE: added goToUserProfile() for ComparatorPresenter & ComparatorInteractor & ComparatorInputBoundary.

FIX: checkstyle and all the javadoc for classes and function are completed.

chrisliux committed:
FIX: Removed unused FunctionState since we are not storing any data in FunctionView, and we don't need to store application state(user's & guest's Function View are the same).

chrisliux committed:
FEATURE: Store the other houses data in ComparatorState; FIX: modified propertyChange method in ComparatorView in order to align to set other houses data from ComparatorState. FEATURE: According to state.isLoggedIn, userprofileButton will be removed if it is a guest using the application (In ComparatorView's propertyChange method).

chrisliux committed:
FIX: Removed unused ComparatorOutputData.java since I moved those into ComparatorState

chrisliux committed:
FEATURE: finished ComparatorInteractorTest & FunctionInteractorTest and reached 100% coverage. FIX: javadoc typo in FunctionFactory.

chrisliux committed:
FIX: Accidentally deleted this file originally from main, restore it here

chrisliux committed:
feat: ApplicationState done by Hadid

samnjab committed:
Merge branch 'main' into christinee.liu

samnjab commented on .idea/compiler.xml:
Make sure .idea is added to .gitignore file so moving forward, the xml files in this folder won't be altered.

samnjab commented on .idea/misc.xml:
Same issue here, the JDK builds are different across branches, make sure you exclude build level files from commits.

samnjab commented on src/main/java/quotify_app/adapters/comparator/ComparatorState.java:
```diff
@@ -0,0 +1,79 @@
+     *
+     * @param list1 the {@code ArrayList} to set as list1. If {@code null}, no changes are made.
+     */
+    public void setList1(ArrayList<Object> list1) {
```
The use of Object as the specified type in method arguments is overly general. It can lead to type-casting issues. Use <Property> instead to make sure type issues are detected at compile time.
Using ArrayList<> in the method signatures restricts flexibility. A user might have a LinkedList or another List implementation they'd like to pass in. Use the List<> interface in both the class fields and method parameters.

samnjab replied:
Use `List<Property>` instead.

chrisliux replied:
You're right. I created a new Property class in entities folder to holds the data for each property.

samnjab commented on src/main/java/quotify_app/adapters/comparator/ComparatorState.java:
```diff
@@ -0,0 +1,79 @@
+ */
+public class ComparatorState {
+    private boolean isLoggedIn;
+    private ArrayList<Object> list1;
```
Use List<Property> for class fields for more flexibility in client code of this class, as well as better readability of the type of objects these fields are supposed to hold.

samnjab replied:
I don't know why parts of my comments get deleted. Use `List<Property>` is what I meant to say

samnjab commented on src/main/java/quotify_app/adapters/comparator/ComparatorState.java:
```diff
@@ -0,0 +1,79 @@
+public class ComparatorState {
+    private boolean isLoggedIn;
+    private ArrayList<Object> list1;
+    private ArrayList<Object> list2;
```
It is not clear what list1, list2, and list3 are representing. Be more descriptive with the field names.

chrisliux replied:
Identified the private attribute **List<Property> properties** in javadoc

samnjab commented on src/main/java/quotify_app/adapters/comparator/ComparatorViewModel.java:
```diff
@@ -0,0 +1,23 @@
+
+    // Labels for UI elements in the login view
+    public static final String TITLE_LABEL = "Property Comparator";
+    public static final String HOUSE1_LABEL = "House 1";
```
Choosing to restrict the number of properties presented in the ViewModel is interesting.  We could discuss whether the decision of how many Properties are presented in the View should be made by the Interactor or the View. 
So question is: are the "similar properties" to be presented better kept in an array, in which case you only need one label for the whole group of them say "Similar Properties", instead of 3 separate labels House 1, House 2, House 3, which could very well be subject to change depending on whether we choose to present more similar properties.

chrisliux replied:
I tried modify not to restrict the number of properties presented in the ViewModel by storing each property list be stored an ArrayList (which can be dynamically adjusts its capacity when elements are added or removed).

samnjab commented on src/main/java/quotify_app/adapters/comparator/ComparatorViewModel.java:
```diff
@@ -0,0 +1,23 @@
+
+    // Labels for UI elements in the login view
+    public static final String TITLE_LABEL = "Property Comparator";
+    public static final String HOUSE1_LABEL = "House 1";
```
One minor note is not all Properties are Houses, some are condos etc.

chrisliux replied:
Nice catch! I changed all the 'house' to 'property' across my code.

samnjab commented on src/main/java/quotify_app/adapters/function/FunctionViewModel.java:
```diff
@@ -0,0 +1,19 @@
+    // Labels for UI elements in the login view
+    public static final String TITLE_LABEL = "Choose!";
+    public static final String ESTIMATE_PRICE_LABEL = "Estimate Current";
+    public static final String COMPARE_LABEL = "Compare with similar property";
```
Minor spelling note in the labels:
Estimate Price
Compare Similar Properties

chrisliux committed:
FIX: javadoc typo and change 'house' to 'property'.
FEATURE: added Property into entities folder to align the ComparatorState of storing properties in an ArrayList. And store each property data as a Property class (which is a map).

chrisliux committed:
FIX: Provided better clarification for the attribute name.

chrisliux committed:
FIX: Change the type of Property class from Map to List<String> to avoid the need to write propertyData.put(key, value) every time writing in data from the api model.

chrisliux committed:
Merge remote-tracking branch 'origin/christinee.liu' into christinee.liu

samnjab reviewed (approved):
Awesome! I'm approving the new Property entity to prevent crashing for now , but keep in mind this is going to be replaced by a more comprehensive entity.
